src/smc-webapp/launch/custom-image.ts

```diff
@@ -3,76 +3,140 @@
+//
+// testing: app?launch=csi/course-calculate-20
+
 import { redux } from "../app-framework";
 import { analytics_event } from "../tracker";
 import { uuid } from "smc-util/misc2";
-import { retry_until_success, once } from "smc-util/async-utils";
+import { retry_until_success } from "smc-util/async-utils";
 import {
   custom_image_name,
   NAME as CUSTOM_SOFTWARE_NAME,
-} from "../custom-software/init";
-
-export async function launch_custom_software_image(
-  launch: string
-): Promise<void> {
-  // processing e.g. "?launch=csi/opencv-machine-learning",
-  // where the ID is a valid docker ID (lowercase, dashes)
-  const image_id = launch.split("/")[1];
-  // console.log(`launching custom software image with ID = ${image_id}`);
-
-  const custom_software_table = await retry_until_success({
-    f: async () => {
-      const cst = redux.getTable(CUSTOM_SOFTWARE_NAME);
-      if (cst == null)
-        throw new Error("custom software table not yet available...");
-      // what is this doing?
-      await once(cst._table, "connected");
-      return cst;
-    },
-  });
-
-  if (!custom_software_table._table.value.has(image_id)) {
-    console.error(
-      `There is no custom software image with the ID "${image_id}"`
-    );
-    return;
+} from "../custom-software/util";
+import { init } from "../custom-software/init";
+import { alert_message } from "../alerts";
+
+export class CSILauncher {
+  private readonly image_id;
+  private actions;
+  private custom_software_table;
+
+  constructor(image_id: string) {
+    // processing e.g. "?launch=csi/opencv-machine-learning",
+    // where the ID is a valid docker ID (lowercase, dashes)
+    this.image_id = image_id;
   }
 
-  // this is mimicing what's going on in projects/create-project.tsx
-  const actions = await retry_until_success({
-    f: async () => {
-      const projects_table = redux.getTable("projects");
-      if (projects_table == null)
-        throw new Error("Projects Table not yet available...");
-      // what is this doing?
-      await once(projects_table._table, "connected");
-      const actions = redux.getActions("projects");
-      if (actions == null)
-        throw new Error("Projects Actions not yet available...");
-      return actions;
-    },
-  });
-
-  const token = uuid();
-
-  // TODO pick the proper title from the custom image table
-  actions.create_project({
-    title: image_id,
-    image: custom_image_name(image_id),
-    token,
-  });
-
-  // if we have project actions, we can assume project store also exists?
-  redux
-    .getStore("projects")
-    .wait_until_project_created(token, 30, (err, project_id) => {
-      if (err != null) {
-        console.error(`Error creating project -- ${err}`);
-      } else {
-        actions.apply_default_upgrades({ project_id });
-        actions.open_project({ project_id, switch_to: true });
-      }
+  // returns project_id or if there is a problem, undefined
+  async launch(): string | undefined {
```

Comment:
@haraldschilly I want to point out that you totally got bit by the ts-fork checker bug here!!   This is clearly invalid code, since an async function can only return a Promise:

![image](https://user-images.githubusercontent.com/1276278/84576558-76d48800-ad6a-11ea-87f5-e551aa6a0e5a.png)

It silently failed to be reported due to ts-forker breaking...  There will be many typescript errors in your code, probably.

I think I finally fixed this problem using a different webpack config option.

Reply:
I'll fix all these for you.